### PR TITLE
Manage server resources with Codensity.

### PR DESCRIPTION
**Pull request by rossabaker**
Provides some helpers and an example of how to gracefully start and stop a server and all its dependencies.

### CODE REVIEW

**Comment by rossabaker** [comment]
I'm not sure this is a good idea yet, but throwing it out there for comment.

**Comment by rossabaker** [comment]
Arguments for:
- Solves a common problem people will have.
- Could be extended to tests as well.

Arguments against:
- Codensity is obscure. I found it by following the types.
- There is no Codensity equivalent in Cats.  This hurts a future shims/cats strategy.
- Not sure how I feel about about the `Task.now` used to acquire in a couple places.

Possible improvements:
- A Manageable typeclass?  (But isn't it lawless?)
- Tie into Java's closeable?
- What else is Manageable in our API?

**Comment by bryce-anderson** [comment]
I don't see anything wrong in the code but it feels a bit out of scope to me. I know there are a few resource management libraries floating around such as scala-arm. I don't know if scala-arm would be useful to us but I'd personally rather leverage existing work than build our own if possible.

**Comment by rossabaker** [comment]
Yes, `util` is shorthand for "this feels out of scope, but it itches."  I don't like having this and I don't like not having this.

Alternatives and buts:
* scala-arm is the best known, but it's better suited to an `Id` ecosystem than a `Task` ecosystem.
* scalaz-effect has a `Resource`, but it doesn't seem to do much.
* this could be a microlibrary, but then how do we integrate client and server?

File: server/src/main/scala/org/http4s/server/ManagedServerApp.scala [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      Codensity.rep(terminateServer(server))
+    }
+
+  /** Prompts for input on the console, and then terminates the server.
+    * Override to implement your own termination condition.
+    */
+  def terminateServer(server: Server): Task[Unit] =
+    Task.async { cb =>
+      cb(\/.fromTryCatchNonFatal {
+        Console.readLine(s"Press <Enter> to terminate server on ${server.address}\n")
```
Comment (rossabaker) [inline server/src/main/scala/org/http4s/server/ManagedServerApp.scala]: Note: changing this to a JVM shutdown hook for ^C doesn't wind down the stack the way that this implementation does.  I think the thread on which the codensity runs already exits.  There might be a way to salvage this with a daemon thread and a lock, but this is where I start to feel seasick.

**Comment by rossabaker** [comment]
Freezing this because I haven't convinced even myself it merits inclusion, but still find it interesting enough to hope passers by discuss it.

**Comment by rossabaker** [comment]
This was a fun experiment, but #595 pretty well scratches this same itch for me.

**Pull request closed without merging**